File: app/src/main/java/com/hyapp/achat/model/People.java
package com.hyapp.achat.model;

public class People extends Thing {

    public static final byte RANK_GUEST = 0;
    public static final byte RANK_MEMBER = 1;
    public static final byte RANK_SPECIAL = 2;
    public static final byte RANK_ACTIVE = 3;
    public static final byte RANK_SENIOR = 4;
    public static final byte RANK_ADMIN = 5;
    public static final byte RANK_MANAGER = 6;

    private Avatar avatar;
    private Key key;

    public People() {
    }

    public People(String name, String bio, byte gender) {
        super(name, bio, gender);
    }

    public People(String name, String bio, byte gender, Avatar avatar) {
        this(name, bio, gender);
        this.avatar = avatar;
    }

    public People(String name, String bio, byte gender, Avatar avatar, Key key) {
        super(name, bio, gender);
        this.avatar = avatar;
        this.key = key;
    }

    public Avatar getAvatar() {
        return avatar;
    }

    public void setAvatar(Avatar avatar) {
        this.avatar = avatar;
    }

    public Key getKey() {
        return key;
    }

    public void setKey(Key key) {
        this.key = key;
    }

    public static int compare(People o1, People o2) {
        if (o1.getKey().getIpv4().equals(o2.getKey().getIpv4())) return 0;
        if (o1.getKey().getRank() < o2.getKey().getRank()) return 1;
        if (o1.getKey().getRank() > o2.getKey().getRank()) return -1;
        if (o1.getKey().getScore() < o2.getKey().getScore()) return 1;
        if (o1.getKey().getScore() > o2.getKey().getScore()) return -1;
        if (o1.getKey().getLoginTime() < o2.getKey().getLoginTime()) return -1;
        if (o1.getKey().getLoginTime() > o2.getKey().getLoginTime()) return 1;
        return 0;
    }
}

File: app/src/main/java/com/hyapp/achat/model/People.kt
package com.hyapp.achat.model

import android.os.Bundle
import com.google.gson.annotations.Expose
import com.hyapp.achat.model.utils.PersonUtils
import io.objectbox.annotation.BaseEntity
import io.objectbox.annotation.Transient

@BaseEntity
open class People : Person {

    @Transient
    var key: Key? = null
        set(value) {
            field = value
            setupRank()
        }

    @Expose
    var avatars: Array<String?> = emptyArray()

    @Transient
    var rankStrRes: Int = PersonUtils.RANK_STR_GUEST

    @Transient
    var rankColor: Int = PersonUtils.RANK_COLOR_GUEST

    constructor(name: String = "",
                bio: String? = null,
                gender: Byte = GENDER_MALE,
                key: Key? = null,
                avatars: Array<String?> = emptyArray()
    ) : super(name, bio, gender) {
        this.key = key
        this.avatars = avatars
    }

    init {
        setupRank()
    }

    constructor(bundle: Bundle) : super(bundle) {
        key = bundle.getParcelable(EXTRA_KEY) ?: Key()
        avatars = bundle.getStringArray(EXTRA_AVATARS) ?: emptyArray()
    }

    override val bundle: Bundle
        get() {
            return super.bundle.apply {
                putParcelable(EXTRA_KEY, key)
                putStringArray(EXTRA_AVATARS, avatars)
            }
        }

    open fun setupRank() {
        key?.let {
            val pair = PersonUtils.rankInt2rankStrResAndColor(it.rank)
            rankStrRes = pair.first
            rankColor = pair.second
        }
    }

    companion object {
        const val EXTRA_KEY = "key"
        const val EXTRA_AVATARS = "avatars"

        @JvmStatic
        fun compare(o1: People, o2: People): Int {
            if (o1.key!!.uid == o2.key!!.uid) return 0
            if (o1.key!!.rank < o2.key!!.rank) return 1
            if (o1.key!!.rank > o2.key!!.rank) return -1
            if (o1.key!!.score < o2.key!!.score) return 1
            if (o1.key!!.score > o2.key!!.score) return -1
            if (o1.key!!.loginTime < o2.key!!.loginTime) return -1
            return if (o1.key!!.loginTime > o2.key!!.loginTime) return 1 else 0
        }
    }
}